File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/bridge/Network.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.reef.bridge;

import org.apache.avro.specific.SpecificRecord;
import org.apache.reef.tang.Injector;
import org.apache.reef.tang.Tang;
import org.apache.reef.wake.EventHandler;
import org.apache.reef.wake.MultiObserver;
import org.apache.reef.wake.avro.ProtocolSerializer;
import org.apache.reef.wake.remote.RemoteMessage;
import org.apache.reef.wake.remote.impl.ByteCodec;
import org.apache.reef.wake.remote.impl.SocketRemoteIdentifier;
import org.apache.reef.wake.remote.RemoteIdentifier;
import org.apache.reef.wake.remote.RemoteManager;
import org.apache.reef.wake.remote.RemoteManagerFactory;
import org.apache.reef.wake.remote.ports.TcpPortProvider;
import org.apache.reef.wake.impl.LoggingEventHandler;
import org.apache.reef.wake.remote.address.LocalAddressProvider;

import javax.inject.Inject;
import java.net.InetSocketAddress;
import java.util.concurrent.atomic.AtomicLong;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * The CLR Bridge Network class aggregates a RemoteManager and
 * Protocol Serializer to provide a simple send/receive interface
 * between the Java and CLR sides of the bridge.
 */
public final class Network {
  private static final Logger LOG = Logger.getLogger(Network.class.getName());
  private static final AtomicLong IDENTIFIER_SOURCE = new AtomicLong(0);

  private RemoteManager remoteManager;
  private InetSocketAddress inetSocketAddress;
  private EventHandler<byte[]> sender;
  private final LocalObserver localObserver;
  private final ProtocolSerializer serializer = new ProtocolSerializer("org.apache.reef.bridge.message");

  /**
   * Sends and receives messages between the java bridge and C# bridge.
   * @param localAddressProvider Local address provider used to find open port.
   * @param observer A multiobserver instance that will receive all incoming messages.
   */
  @Inject
  public Network(final LocalAddressProvider localAddressProvider, final MultiObserver observer) {
    LOG.log(Level.INFO, "Initializing");

    this.localObserver = new LocalObserver(this, observer);

    try {
      String name = "JavaBridgeNetwork";
      int port = 0;
      boolean order = true;
      int retries = 3;
      int timeOut = 10000;

      // Instantiate a port provider.
      final Injector injector = Tang.Factory.getTang().newInjector();
      final TcpPortProvider tcpPortProvider = injector.getInstance(TcpPortProvider.class);

      // Instantiate a remote manager to handle java-C# bridge communication.
      final RemoteManagerFactory remoteManagerFactory = injector.getInstance(RemoteManagerFactory.class);
      remoteManager = remoteManagerFactory.getInstance(
        name, localAddressProvider.getLocalAddress(), port, new ByteCodec(),
        new LoggingEventHandler<Throwable>(), order, retries, timeOut,
        localAddressProvider, tcpPortProvider);

      // Get our address and port number.
      final RemoteIdentifier remoteIdentifier = remoteManager.getMyIdentifier();
      if (remoteIdentifier instanceof SocketRemoteIdentifier) {
        SocketRemoteIdentifier socketIdentifier = (SocketRemoteIdentifier)remoteIdentifier;
        inetSocketAddress = socketIdentifier.getSocketAddress();
      } else {
        throw new RuntimeException("Identifier is not a SocketRemoteIdentifier");
      }

      // Register as the message handler for any incoming messages.
      remoteManager.registerHandler(byte[].class, this.localObserver);

    } catch (final Exception e) {
      LOG.log(Level.SEVERE, "Initialization failed: ", e);
    }
  }

  /**
   * Sends a message to the C# side of the bridge.
   * @param message An Avro message class derived from SpecificRecord.
   */
  public void send(final long identifier, final SpecificRecord message) {
    if (sender != null) {
      sender.onNext(serializer.write(message, identifier));
    } else {
      LOG.log(Level.SEVERE,
          "Attempt to send message [{0}] before network is initialized", message.getClass().getName());
    }
  }

  /**
   * Provides the IP address and port of the java bridge network.
   * @return A InetSockerAddress that contains the ip and port of the bridge network.
   */
  public InetSocketAddress getAddress() {
    return inetSocketAddress;
  }

  /**
   * Processes messages from the network remote manager.
   */
  private class LocalObserver implements EventHandler<RemoteMessage<byte[]>> {
    private final Network network;
    private final MultiObserver messageObserver;

    /**
     * Associate the local observer with the specified network and message observer.
     * @param network
     * @param messageObserver
     */
    LocalObserver(final Network network, final MultiObserver messageObserver) {
      this.network = network;
      this.messageObserver = messageObserver;
    }

    /**
     * Deserialize and direct incoming messages to the registered MuiltiObserver event handler.
     * @param message A RemoteMessage<byte[]> object which will be deserialized.
     */
    public void onNext(final RemoteMessage<byte[]> message) {
      LOG.log(Level.INFO, "Received remote message: {0}", message);

      if (network.sender == null) {
        // Instantiate a network connection to the C# side of the bridge.
        // THERE COULD BE  A SECURITY ISSUE HERE WHERE SOMEONE SPOOFS THE
        // C# BRIDGE, WE RECEIVE IT FIRST, AND CONNECT TO THE SPOOFER,
        // THOUGH THE TIME WINDOW IS VERY SMALL.
        final RemoteIdentifier remoteIdentifier = message.getIdentifier();
        LOG.log(Level.INFO, "Connecting to: {0}", remoteIdentifier);
        network.sender = remoteManager.getHandler(remoteIdentifier, byte[].class);
      }

      // Deserialize the message and invoke the appropriate processing method.
      serializer.read(message.getMessage(), messageObserver);
    }
  }
}

File: lang/cs/Org.Apache.REEF.Bridge.CLR/Network.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System;
using System.Collections.Concurrent;
using System.IO;
using System.Net;
using org.apache.reef.bridge.message;
using Org.Apache.REEF.Common.Files;
using Org.Apache.REEF.Tang.Annotations;
using Org.Apache.REEF.Tang.Formats;
using Org.Apache.REEF.Tang.Implementations.Tang;
using Org.Apache.REEF.Tang.Util;
using Org.Apache.REEF.Utilities.Logging;
using Org.Apache.REEF.Wake.Avro;
using Org.Apache.REEF.Wake.Remote;
using Org.Apache.REEF.Wake.Remote.Impl;

namespace Org.Apache.REEF.Bridge
{
    public sealed class NetworkOptions
    {
        [NamedParameter(Documentation = "Message observer bridge Avro messages")]
        public class MessageObserver : Name<object>
        {
        }

        public sealed class ModuleBuilder : ConfigurationModuleBuilder
        {
            public static readonly RequiredParameter<object> MessageObserver = new RequiredParameter<object>();

            public static readonly ConfigurationModule Config = new ModuleBuilder()
                .BindNamedParameter(GenericType<NetworkOptions.MessageObserver>.Class, MessageObserver)
                .Build();
        }
    }

    /// <summary>
    /// The CLR Bridge Network class agregates a RemoteManager and
    /// Protocol Serializer to provide a simple send/receive interface
    /// between the CLR and Java bridges. 
    /// </summary>
    public sealed class Network
    {
        private static readonly Logger Logger = Logger.GetLogger(typeof(Network));
        private readonly ProtocolSerializer serializer =
            new ProtocolSerializer(typeof(Network).Assembly, "org.apache.reef.bridge.message");
        private readonly BlockingCollection<byte[]> queue = new BlockingCollection<byte[]>();
        private readonly IRemoteManager<byte[]> remoteManager;
        private readonly IObserver<byte[]> remoteObserver;
        private readonly LocalObserver localObserver;

        /// <summary>
        /// Construct a network stack using the wate remote manager.
        /// </summary>
        /// <param name="localAddressProvider">An address provider used to obtain a local IP address on an open port.</param>
        /// <param name="messageObserver">Message receiver that implements IObserver interface for each message to be processed</param>
        [Inject]
        public Network(
            ILocalAddressProvider localAddressProvider,
            [Parameter(typeof(NetworkOptions.MessageObserver))] object messageObserver)
        {
            this.localObserver = new LocalObserver(serializer, messageObserver);

            // Get the path to the bridge name server endpoint file.
            string javaBridgeAddress = GetJavaBridgeAddress();

            // Instantiate the remote manager.
            IRemoteManagerFactory remoteManagerFactory =
                TangFactory.GetTang().NewInjector().GetInstance<IRemoteManagerFactory>();
            remoteManager = remoteManagerFactory.GetInstance(localAddressProvider.LocalAddress, new ByteCodec());

            // Listen to the java bridge on the local end point.
            remoteManager.RegisterObserver(localObserver);
            Logger.Log(Level.Info, "Local observer listening to java bridge on: [{0}]", remoteManager.LocalEndpoint);

            // Instantiate a remote observer to send messages to the java bridge.
            string[] javaAddressStrs = javaBridgeAddress.Split(':');
            IPAddress javaBridgeIpAddress = IPAddress.Parse(javaAddressStrs[0]);
            int port = int.Parse(javaAddressStrs[1]);
            IPEndPoint javaIpEndPoint = new IPEndPoint(javaBridgeIpAddress, port);
            remoteObserver = remoteManager.GetRemoteObserver(javaIpEndPoint);
            Logger.Log(Level.Info, "Connecting to java bridge on: [{0}]", javaIpEndPoint);

            // Negotiate the protocol.
            Send(0, new BridgeProtocol(100));
        }

        /// <summary>
        /// Send a message to the java side of the bridge.
        /// </summary>
        /// <param name="identifier">A long value that which is the unique sequence identifier of the message.</param>
        /// <param name="message">An object reference to a message in the org.apache.reef.bridge.message package.</param>
        public void Send(long identifier, object message)
        {
            Logger.Log(Level.Info, "Sending message: {0}", message);
            remoteObserver.OnNext(serializer.Write(message, identifier));
        }

        /// <summary>
        /// The Local Observer class receives byte buffer messages from the transport layer,
        /// deserializes the messages into Avro C# classes, and invokes the appropriate
        /// IObserver callback on the Avro message observer.
        /// </summary>
        private class LocalObserver : IObserver<IRemoteMessage<byte[]>>
        {
            private readonly ProtocolSerializer serializer;
            private readonly object messageObserver;
            [Inject]
            public LocalObserver(ProtocolSerializer serializer, object messageObserver)
            {
                this.serializer = serializer;
                this.messageObserver = messageObserver;
            }

            /// <summary>
            /// Called by the remote manager to process messages received from the java bridge.
            /// </summary>
            /// <param name="message">A byte buffer containing a serialied message.</param>
            public void OnNext(IRemoteMessage<byte[]> message)
            {
                Logger.Log(Level.Info, "Message received: {0}", message.Identifier);

                // Deserialize the message and invoke the appropriate handler.
                serializer.Read(message.Message, messageObserver);
            }

            /// <summary>
            /// Handles error conditions in the low transport layer.
            /// </summary>
            /// <param name="error">The exception generated in the transport layer.</param>
            public void OnError(Exception error)
            {
                Logger.Log(Level.Info, "Error: [{0}]", error.Message);
            }

            /// <summary>
            /// Notification that no nore message prpocessing is required.
            /// </summary>
            public void OnCompleted()
            {
                Logger.Log(Level.Info, "Completed");
            }
        }

        /// <summary>
        /// Retrieves the address of the java bridge.
        /// </summary>
        /// <returns>A string containing the IP address and port of the Java bridge.</returns>
        private string GetJavaBridgeAddress()
        {
            string javaBridgeAddress = null;
            REEFFileNames fileNames = TangFactory.GetTang().NewInjector().GetInstance<REEFFileNames>();
            using (FileStream stream = File.Open(fileNames.GetDriverJavaBridgeEndpoint(), FileMode.Open))
            {
                using (StreamReader reader = new StreamReader(stream))
                {
                    javaBridgeAddress = reader.ReadToEnd();
                }
            }
            Logger.Log(Level.Info, "Name Server Address: {0}", javaBridgeAddress);
            return javaBridgeAddress;
        }
    }
}